### ISSUE TITLE
Adding JamtabaVST2 after deleting it fails

### ISSUE THREAD

**Issue by pljones** [elieserdejesus/JamTaba#427]
I was testing my changes to Jamtaba2 and thought I'd broken something but it turns out the same failure happens in the current release version.

Add JamtabaVST2 to a MuLab Rack slot
Delete the JamtabaVST2 entry
Add it again

You get a "Couldn't create VST instance for JamtabaVST2" error message.

I've not looked into why or whether it's specific to MuLab.

**Comment by pljones** [comment]
A little further investigation.

Turning all logging on reveals that, even after deleting the plugin from the rack, it's still running:
```
jt.VstPlugin.DEBUG:  VstEditor::close()...  in Editor.cpp 107
jt.VstPlugin.DEBUG:  VstEditor::close() done.  in Editor.cpp 116
jt.VstPlugin.DEBUG:  JamtabaPLugin::suspend()  in Plugin.cpp 265
jt.VstPlugin.DEBUG:  JamtabaPLugin::resume()  in Plugin.cpp 270 <-----------
jt.IpToLocation.DEBUG:  cache hit for  "178.4.71.128"  in WebIpToLocationResolver.cpp 169
jt.IpToLocation.DEBUG:  cache hit for  "84.245.28.128"  in WebIpToLocationResolver.cpp 169
jt.IpToLocation.DEBUG:  cache hit for  "82.249.136.128"  in WebIpToLocationResolver.cpp 169
jt.IpToLocation.DEBUG:  cache hit for  "82.242.123.128"  in WebIpToLocationResolver.cpp 169
jt.IpToLocation.DEBUG:  cache hit for  "178.4.71.128"  in WebIpToLocationResolver.cpp 169
jt.IpToLocation.DEBUG:  cache hit for  "84.245.28.128"  in WebIpToLocationResolver.cpp 169
jt.IpToLocation.DEBUG:  cache hit for  "82.249.136.128"  in WebIpToLocationResolver.cpp 169
jt.IpToLocation.DEBUG:  cache hit for  "82.242.123.128"  in WebIpToLocationResolver.cpp 169
jt.IpToLocation.DEBUG:  cache hit for  "178.4.71.128"  in WebIpToLocationResolver.cpp 169
jt.IpToLocation.DEBUG:  cache hit for  "84.245.28.128"  in WebIpToLocationResolver.cpp 169
jt.IpToLocation.DEBUG:  cache hit for  "82.249.136.128"  in WebIpToLocationResolver.cpp 169
jt.IpToLocation.DEBUG:  cache hit for  "82.242.123.128"  in WebIpToLocationResolver.cpp 169
```
`<-----------` indicates roughly where I deleted the plugin - the log should stop....

**Comment by pljones** [comment]
And a little more.

When I finally close the project file, the log shows
```
jt.IpToLocation.DEBUG:  cache hit for  "178.4.71.128"  in WebIpToLocationResolver.cpp 169
jt.IpToLocation.DEBUG:  cache hit for  "84.245.28.128"  in WebIpToLocationResolver.cpp 169
jt.IpToLocation.DEBUG:  cache hit for  "82.249.136.128"  in WebIpToLocationResolver.cpp 169
jt.IpToLocation.DEBUG:  cache hit for  "82.242.123.128"  in WebIpToLocationResolver.cpp 169
jt.VstPlugin.DEBUG:  JamtabaPLugin::suspend()  in Plugin.cpp 265
jt.VstPlugin.DEBUG:  JamtabaPlugin::close()  in Plugin.cpp 139
jt.VstPlugin.DEBUG:  JamtabaPLugin::close() finished  in Plugin.cpp 147
jt.VstPlugin.DEBUG:  Plugin destructor  in Plugin.cpp 104
jt.Core.DEBUG:  MainController destrutor!  in MainController.cpp 537
jt.Core.DEBUG:  Stopping MainController...  in MainController.cpp 722
jt.Core.DEBUG:  disconnecting from login server...  in MainController.cpp 729
jt.LoginService.DEBUG:  DISCONNECTING from server...  in LoginService.cpp 176
jt.LoginService.DEBUG:  sending disconnect command to server...  in LoginService.cpp 181
jt.GUI.DEBUG:  MainWindow destructor...  in MainWindow.cpp 931
jt.GUI.DEBUG:  Main frame timer killed!  in MainWindow.cpp 941
jt.GUI.DEBUG:  MainWindow destructor finished.  in MainWindow.cpp 942
jt.LoginService.DEBUG:  disconnected  in LoginService.cpp 183
jt.LoginService.DEBUG:  reply content readed and discarded!  in LoginService.cpp 187
jt.LoginService.DEBUG:  disconnected from login server!  in LoginService.cpp 194
jt.Core.DEBUG:  main controller stopped!  in MainController.cpp 542
jt.Core.DEBUG:  cleaning tracksNodes...  in MainController.cpp 544
jt.Core.DEBUG:  cleaning tracksNodes done!  in MainController.cpp 553
jt.Core.DEBUG:  MainController destructor finished!  in MainController.cpp 555
jt.Cache.DEBUG:  Saving cache file  in UsersDataCache.cpp 179
jt.Cache.DEBUG:  Tracks cache file opened to write.  in UsersDataCache.cpp 183
jt.Cache.DEBUG:  1  items stored in tracks cache file!  in UsersDataCache.cpp 191
jt.IpToLocation.DEBUG:  53  country codes stored in  "country_codes_cache.bin"  in WebIpToLocationResolver.cpp 63
jt.IpToLocation.DEBUG:  16  country names stored in  "country_names_cache_en.bin"  in WebIpToLocationResolver.cpp 54
jt.IpToLocation.DEBUG:  IpToLocationResolver destructor  in IpToLocationResolver.cpp 42
jt.Audio.DEBUG:  Audio mixer destructor...  in AudioMixer.cpp 34
jt.Audio.DEBUG:  Audio mixer destructor finished!  in AudioMixer.cpp 37
jt.LoginService.DEBUG:  LoginService Destructor  in LoginService.cpp 142
jt.LoginService.DEBUG:  LoginService Destructor  in LoginService.cpp 142
jt.VstPlugin.DEBUG:  VstEditor detructor...  in Editor.cpp 29
jt.VstPlugin.DEBUG:  VstEditor detructor done.  in Editor.cpp 31
```
Which is it cleaning itself up.

```
jt.VstPlugin.DEBUG:  Plugin destructor  in Plugin.cpp 104
```
I can see that it's waiting for the host to call its destructor before cleaning up.  However, despite the start up being bulky, if the plugin isn't handling being inserted multiple times *cleanly* (i.e. it fails), then it's better to do the clean up as part of the `suspend()`.

Unfortunately, nothing is logged when I make the second attempt (i.e. after the `suspend()`/`resume()` caused by deleting the plugin) - only the host saying the create instance failed.

**Comment by pljones** [comment]
Checking in Reaper -- it actually calls the destructor when you remove the plugin from a rack, so removing and adding again is like closing and opening a project in MuLab.

I've only the two VST hosts to check in, so I can't comment on which behaviour is more prevalent.

**Comment by elieserdejesus** [comment]
@pljones 
> do the clean up as part of the suspend()

All hosts I tested (Sonar, Reaper, Studio One, Cubase) are suspending plugins correctly. Change this to work in Mulab will broke Jamtaba to everybody.

> I've only the two VST hosts to check in, so I can't comment on which behaviour is more prevalent.

I tested in the mentioned hosts and is working ok. So I'm shure Mulab behavior is the exception. At moment I'm using a static variable to store the Jamtaba instance and detect if users are trying insert 2 Jamtaba instances. I called this a "troll detector flag" :) Maybe this is causing the problem in Mulab, but it's just a guess.

@pljones , I nice trick to inspect plugins is debug the plugin. Put a breakpoint in plugin code and "start debugging using an external application (the host)". I used this trick a lot to solve VST plugin problems:
![image](https://cloud.githubusercontent.com/assets/1012741/15805614/5cdcb7aa-2b06-11e6-99bc-2e81813fad2c.png)
